packages/engine-core/src/library/ExitHooks.ts

```diff
@@ -0,0 +1,67 @@
+import Debug from '@prisma/debug'
+
+export type BeforeExitListener = () => unknown
+
+const debug = Debug('prisma:client:libraryEngine:exitHooks')
+
+export class ExitHooks {
+  private nextOwnerId = 1
+  private ownerToIdMap = new WeakMap<object, number>()
+  private idToListenerMap = new Map<number, BeforeExitListener>()
+  private areHooksInstalled = false
+
+  install() {
+    if (this.areHooksInstalled) {
+      return
+    }
+
+    this.installHook('beforeExit')
+    this.installHook('exit')
+    this.installHook('SIGINT', true)
+    this.installHook('SIGUSR2', true)
+    this.installHook('SIGTERM', true)
+    this.areHooksInstalled = true
+  }
+
+  setListener(owner: object, listener: BeforeExitListener | undefined) {
+    if (listener) {
+      let id = this.ownerToIdMap.get(owner)
+      if (!id) {
+        id = this.nextOwnerId++
+        this.ownerToIdMap.set(owner, id)
+      }
+      this.idToListenerMap.set(id, listener)
+    } else {
+      const id = this.ownerToIdMap.get(owner)
+      if (id !== undefined) {
+        this.ownerToIdMap.delete(owner)
+        this.idToListenerMap.delete(id)
+      }
+    }
+  }
+
+  getListener(owner: object): BeforeExitListener | undefined {
+    const id = this.ownerToIdMap.get(owner)
+    if (id === undefined) {
+      return undefined
+    }
+    return this.idToListenerMap.get(id)
+  }
+
+  private installHook(event: string, shouldExit = false) {
+    process.once(event, async (code) => {
+      debug(`exit event received: ${event}`)
+      for (const listener of this.idToListenerMap.values()) {
+        await listener()
+      }
```

Comment:
`BeforeExitListener` has type `() => unknown`, I think we should explicitly type it as `() => Promise<unknown>` before `await`-ing on `listener()`. Unless both sync and async listeners are supported

Reply:
By the way I don't think it's really possible to run async code in `process.on('exit', ...)` hook. `SIGINT` and `SIGTERM` hooks will remove the default handlers so Node.js will only terminate when we explicitly call `process.exit` after setting them, with `'exit'` Node.js will terminate as soon as all listeners finish (and this listener actually finishes before the first `await` as far as its caller is concerned since it doesn't expect an async function).

Reply:
You're right, `exit` doesn't support asynchronous handlers, as the event loop would've been emptied and abandoned already. If we manually set a listener on `beforeExit` (which we do), we can invoke asynchronous handlers instead (they would be queued to the empty event loop, processed by Node.js, and eventually the `exit` event would be emited)

Reply:
And since `BeforeExitListener` refers to `process.once('beforeExit', fn)`, where `fn: BeforeExitListener`, I guess the right type would be something like:

```typescript
type BeforeExitListener<T = unknown>: () => (Promise<T> | T)
```

Reply:
I don't think we actually need return value for this, so the proper type I beleive is `() => void | Promise<void>` and that's what I changed it to